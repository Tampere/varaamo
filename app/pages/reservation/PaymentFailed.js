import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Button } from 'react-bootstrap';

import injectT from '../../i18n/injectT';

function PaymentFailed({
  t,
  resourceId,
}) {
  return (
    <div className="app-ReservationFailedPage__contentbox">
      <h2>{t('payment.title')}</h2>
      <p>{t('payment.text')}</p>
      {resourceId && (
        <Link className="app-ReservationFailedPage__link" to={`/resources/${resourceId}`}>
          <Button bsStyle="primary">{t('payment.return')}</Button>
        </Link>
      )}
      <Link className="app-ReservationFailedPage__link app-ReservationFailedPage__link--last" to="/my-reservations">
        <Button bsStyle="primary">{t('payment.link')}</Button>
      </Link>
    </div>
  );
}

PaymentFailed.propTypes = {
  t: PropTypes.func,
  resourceId: PropTypes.string,
};

export default injectT(PaymentFailed);
