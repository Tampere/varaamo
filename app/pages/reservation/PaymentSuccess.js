import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { injectIntl } from 'react-intl';
import flow from 'lodash/flow';
import get from 'lodash/get';
import isPlainObject from 'lodash/isPlainObject';

import { currentUserSelector } from '../../state/selectors/authSelectors';
import ReservationConfirmation from './reservation-confirmation/ReservationConfirmation';

const translateEntity = (entity, locale) => (
  Object
    .entries(entity)
    .reduce((acc, [key, value]) => {
      const localizedValue = get(value, locale, value);
      acc[key] = isPlainObject(localizedValue) ? null : localizedValue;
      return acc;
    }, {})
);

function PaymentSuccess({
  reservation,
  resource,
  user,
  intl: { locale },
}) {
  const translatedReservation = translateEntity(reservation, locale);
  const translatedResource = translateEntity(resource, locale);
  return (
    <div>
      <ReservationConfirmation
        reservation={translatedReservation}
        resource={translatedResource}
        user={user}
      />
    </div>
  );
}

PaymentSuccess.propTypes = {
  reservation: PropTypes.object,
  resource: PropTypes.object,
  user: PropTypes.object,
  intl: PropTypes.object,
};

export default flow(
  connect(
    state => ({
      user: currentUserSelector(state),
    }),
  ),
  injectIntl,
)(PaymentSuccess);
