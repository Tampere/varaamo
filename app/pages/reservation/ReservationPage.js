import first from 'lodash/first';
import isEmpty from 'lodash/isEmpty';
import last from 'lodash/last';
import get from 'lodash/get';
import has from 'lodash/has';
import moment from 'moment';
import React, { Component } from 'react';
import Loader from 'react-loader';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import queryString from 'query-string';
import { decamelizeKeys } from 'humps';

import { addNotification } from '../../actions/notificationsActions';
import { postReservation, putReservation } from '../../actions/reservationActions';
import { fetchResource } from '../../actions/resourceActions';
import {
  clearReservations,
  closeReservationSuccessModal,
  setSelectedTimeSlots,
} from '../../actions/uiActions';
import PageWrapper from '../PageWrapper';
import injectT from '../../i18n/injectT';
import ReservationConfirmation from './reservation-confirmation/ReservationConfirmation';
import ReservationInformation from './reservation-information/ReservationInformation';
import ReservationPhases from './reservation-phases/ReservationPhases';
import ReservationTime from './reservation-time/ReservationTime';
import reservationPageSelector from './reservationPageSelector';
import { hasProducts } from '../../utils/resourceUtils';
import RecurringReservationControls from '../../shared/recurring-reservation-controls/RecurringReservationControls';
import CompactReservationList from '../../shared/compact-reservation-list/CompactReservationList';
import recurringReservationsConnector from '../../state/recurringReservations';

class UnconnectedReservationPage extends Component {
  constructor(props) {
    super(props);
    this.fetchResource = this.fetchResource.bind(this);
    const { reservationToEdit } = this.props;
    this.state = {
      view: !isEmpty(reservationToEdit) ? 'time' : 'information',
    };
  }

  componentDidMount() {
    const {
      location,
      reservationCreated,
      reservationEdited,
      reservationToEdit,
      selected,
      history,
    } = this.props;
    if (
      isEmpty(reservationCreated)
      && isEmpty(reservationEdited)
      && isEmpty(reservationToEdit)
      && isEmpty(selected)
    ) {
      const query = queryString.parse(location.search);
      if (!query.id && query.resource) {
        history.replace(`/resources/${query.resource}`);
      } else {
        history.replace(query.path ? '/manage-reservations' : '/my-reservations');
      }
    } else {
      this.fetchResource();
      window.scrollTo(0, 0);
    }
  }

  componentWillUpdate(nextProps) {
    const { reservationCreated: nextCreated, reservationEdited: nextEdited } = nextProps;
    const { reservationCreated, reservationEdited } = this.props;
    if (
      (!isEmpty(nextCreated) || !isEmpty(nextEdited))
      && (nextCreated !== reservationCreated || nextEdited !== reservationEdited)
    ) {
      // Reservation created for resource with product/order: proceed to payment!
      if (has(nextCreated, 'order.paymentUrl')) {
        const paymentUrl = get(nextCreated, 'order.paymentUrl');
        window.location = paymentUrl;
        return;
      }
      // eslint-disable-next-line react/no-will-update-set-state
      this.setState({ view: 'confirmation' });
      window.scrollTo(0, 0);
    }
  }

  componentWillUnmount() {
    this.props.actions.clearReservations();
    this.props.actions.closeReservationSuccessModal();
  }

  handleBack = () => {
    if (!isEmpty(this.props.reservationToEdit)) {
      this.setState({ view: 'time' });
      window.scrollTo(0, 0);
    }
  };

  handleCancel = () => {
    const {
      reservationToEdit, resource, history, location,
    } = this.props;
    if (!isEmpty(reservationToEdit)) {
      const query = queryString.parse(location.search);
      history.replace(query.path ? '/manage-reservations' : '/my-reservations');
    } else {
      history.replace(`/resources/${resource.id}`);
    }
  };

  handleConfirmTime = () => {
    this.setState({ view: 'information' });
    window.scrollTo(0, 0);
  };

  createPaymentReturnUrl = () => {
    const { protocol, hostname } = window.location;
    const port = window.location.port ? `:${window.location.port}` : '';
    return `${protocol}//${hostname}${port}/reservation-payment-return`;
  };

  handleReservation = (values = {}) => {
    const {
      actions, reservationToEdit, resource, selected, recurringReservations = [],
    } = this.props;
    if (!isEmpty(selected)) {
      const { begin } = first(selected);
      const { end } = last(selected);

      if (!isEmpty(reservationToEdit)) {
        const reservation = Object.assign({}, reservationToEdit);
        const decamelizeValues = decamelizeKeys(Object.assign({}, values));
        actions.putReservation({
          ...reservation,
          ...decamelizeValues,
          begin,
          end,
        });
      } else {
        const allReservations = [...recurringReservations, { begin, end }];

        const isOrder = hasProducts(resource);
        const order = isOrder
          ? {
            order: {
              order_lines: [{
                product: get(resource, 'products[0].id'),
              }],
              return_url: this.createPaymentReturnUrl(),
            },
          } : {};

        if (isOrder) {
          this.setState({ view: 'payment' });
        }
        allReservations.forEach(reservation => actions.postReservation({
          ...values,
          ...order,
          begin: reservation.begin,
          end: reservation.end,
          resource: resource.id,
        }));
      }
    }
  };

  fetchResource() {
    const {
      actions, date, resource, location,
    } = this.props;

    const start = moment(date)
      .subtract(1, 'M')
      .startOf('month')
      .toISOString();
    const end = moment(date)
      .add(1, 'M')
      .endOf('month')
      .toISOString();
    const params = queryString.parse(location.search);

    if (!isEmpty(resource)) {
      actions.fetchResource(resource.id, { start, end });
    } else if (params.resource) {
      actions.fetchResource(params.resource, { start, end });
      // Fetch resource by id if there are resource id exist in query but not in redux.
      // TODO: Always invoke actually API call for fetching resource by ID, will fix later.
    }
  }

  renderRecurringReservations = () => {
    const {
      resource,
      actions,
      recurringReservations,
      selectedReservations,
      t,
      isStaff,
    } = this.props;

    const reservationsCount = selectedReservations.length + recurringReservations.length;
    const introText = resource.needManualConfirmation
      ? t('ConfirmReservationModal.preliminaryReservationText', { reservationsCount })
      : t('ConfirmReservationModal.regularReservationText', { reservationsCount });

    return (
      isStaff
        ? (
          <>
            {/* Recurring selection dropdown  */}
            <RecurringReservationControls />
            {<p><strong>{introText}</strong></p>}

            {/* Selected recurring info */}
            <CompactReservationList
              onRemoveClick={actions.removeReservation}
              removableReservations={recurringReservations}
              reservations={selectedReservations}
            />
          </>
        ) : ''
    );
  }

  render() {
    const {
      actions,
      isAdmin,
      isStaff,
      isFetchingResource,
      isMakingReservations,
      location,
      reservationCreated,
      reservationEdited,
      reservationToEdit,
      resource,
      selected,
      t,
      unit,
      user,
      history,
      failedReservations,
      date,
    } = this.props;
    const { view } = this.state;

    if (
      isEmpty(resource)
      && isEmpty(reservationCreated)
      && isEmpty(reservationEdited)
      && isEmpty(reservationToEdit)
      && !isFetchingResource
    ) {
      return <div />;
    }

    const isEditing = !isEmpty(reservationToEdit);
    const isEdited = !isEmpty(reservationEdited);
    const begin = !isEmpty(selected) ? first(selected).begin : null;
    const end = !isEmpty(selected) ? last(selected).end : null;
    const selectedTime = begin && end ? { begin, end } : null;
    const title = t(
      `ReservationPage.${isEditing || isEdited ? 'editReservationTitle' : 'newReservationTitle'}`,
    );

    return (
      <div className="app-ReservationPage">
        <PageWrapper title={title} transparent>
          <div>
            <div className="app-ReservationPage__content">
              <h1 className="app-ReservationPage__title app-ReservationPage__title--big">
                {title}
              </h1>
              <Loader loaded={!isEmpty(resource)}>
                <ReservationPhases
                  currentPhase={view}
                  isEditing={isEditing || isEdited}
                  resource={resource}
                />
                {view === 'time' && isEditing && (
                  <ReservationTime
                    addNotification={actions.addNotification}
                    date={date}
                    handleSelectReservation={actions.setSelectedTimeSlots}
                    history={history}
                    isStaff={isStaff}
                    onCancel={this.handleCancel}
                    onConfirm={this.handleConfirmTime}
                    resource={resource}
                    selectedReservation={reservationToEdit}
                    unit={unit}
                  />
                )}
                {view === 'information' && selectedTime && (
                  <>
                    {isAdmin && this.renderRecurringReservations()}
                    <ReservationInformation
                      isAdmin={isAdmin}
                      isEditing={isEditing}
                      isMakingReservations={isMakingReservations}
                      isStaff={isStaff}
                      onBack={this.handleBack}
                      onCancel={this.handleCancel}
                      onConfirm={this.handleReservation}
                      reservation={reservationToEdit}
                      resource={resource}
                      selectedTime={selectedTime}
                      unit={unit}
                    />
                  </>
                )}
                {view === 'payment' && (
                  <div className="text-center">
                    <p>{t('ReservationPage.paymentText')}</p>
                  </div>
                )}
                {view === 'confirmation' && (reservationCreated || reservationEdited) && (
                  <ReservationConfirmation
                    failedReservations={failedReservations}
                    isEdited={isEdited}
                    location={location}
                    reservation={reservationCreated || reservationEdited}
                    resource={resource}
                    user={user}
                  />
                )}
              </Loader>
            </div>
          </div>
        </PageWrapper>
      </div>
    );
  }
}

UnconnectedReservationPage.propTypes = {
  actions: PropTypes.object.isRequired,
  date: PropTypes.string.isRequired,
  isAdmin: PropTypes.bool.isRequired,
  isStaff: PropTypes.bool.isRequired,
  isFetchingResource: PropTypes.bool.isRequired,
  isMakingReservations: PropTypes.bool.isRequired,
  location: PropTypes.object.isRequired,
  reservationToEdit: PropTypes.object,
  reservationCreated: PropTypes.object,
  reservationEdited: PropTypes.object,
  resource: PropTypes.object.isRequired,
  selected: PropTypes.array.isRequired,
  failedReservations: PropTypes.array.isRequired,
  t: PropTypes.func.isRequired,
  unit: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
  history: PropTypes.object.isRequired,
  recurringReservations: PropTypes.array.isRequired,
  selectedReservations: PropTypes.array.isRequired,
};
UnconnectedReservationPage = injectT(UnconnectedReservationPage); // eslint-disable-line

function mapDispatchToProps(dispatch) {
  const actionCreators = {
    clearReservations,
    closeReservationSuccessModal,
    fetchResource,
    putReservation,
    postReservation,
    removeReservation: recurringReservationsConnector.removeReservation,
    setSelectedTimeSlots,
    addNotification,
  };

  return { actions: bindActionCreators(actionCreators, dispatch) };
}

export { UnconnectedReservationPage };
export default connect(
  reservationPageSelector,
  mapDispatchToProps,
)(UnconnectedReservationPage);
