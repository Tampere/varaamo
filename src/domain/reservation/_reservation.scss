@import 'modal/reservationInformationModal';

@import 'manage/page/manageReservationsPage';
@import 'manage/filters/manageReservationsFilters';
@import 'manage/list/manageReservationsList';
@import 'manage/status/manageReservationsStatus';
