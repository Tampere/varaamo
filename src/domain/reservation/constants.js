// Extracted from Respa Admin, under reservation unit metadata fields
export const RESERVATION_METADATA = [
  'billing_address_city',
  'billing_address_street',
  'billing_address_zip',
  'company',
  'event_description',
  // 'event_description_guidance', <== This is legacy field, wont be used in further development of respa/varaamo
  'event_subject',
  'number_of_participants',
  'reservation_extra_questions',
  'reserver_address_city',
  'reserver_address_street',
  'reserver_address_zip',
  'reserver_email_address',
  'reserver_id',
  'reserver_name',
  'reserver_phone_number',
];

export const RESERVATION_SHOWONLY_FILTERS = {
  FAVORITE: 'favorite',
  CAN_MODIFY: 'can_modify',
};

export const RESERVATION_TYPE = {
  NORMAL: 'normal',
  BLOCKED: 'blocked',
};
