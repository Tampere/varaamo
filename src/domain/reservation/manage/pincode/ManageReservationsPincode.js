import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import { isAccessCodeGenerated, isAccessCodePending } from '../../../../../app/shared/reservation-access-code/helpers';
import TooltipOverlay from '../../../../common/tooltip/TooltipOverlay';
import iconClock from '../../../../../app/assets/icons/clock-o.svg';
import PopoverOverlay from '../../../../common/popover/PopoverOverlay';

class ManageReservationsPincode extends Component {
  renderPincodeField() {
    const { reservation } = this.props;
    if (isAccessCodeGenerated(reservation)) {
      return (
        <TooltipOverlay
          content={(
            <p>{reservation.accessCode}</p>
          )}
        >
          <span>****</span>
        </TooltipOverlay>
      );
    }
    if (isAccessCodePending(reservation, reservation.resource)) {
      return (
        <PopoverOverlay
          content={<FormattedMessage id="ReservationAccessCode.pending" />}
        >
          <img alt="reservationAccessCodePending" src={iconClock} />
        </PopoverOverlay>
      );
    }

    return '';
  }

  render() {
    return (
      <div className="app-ManageReservationPincode">
        {this.renderPincodeField()}
      </div>
    );
  }
}

ManageReservationsPincode.propTypes = {
  reservation: PropTypes.object.isRequired,
};

export default ManageReservationsPincode;
