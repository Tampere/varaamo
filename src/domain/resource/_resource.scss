@import './card/resourceCard';
@import './card/slider/resourceCardSlider';
@import './map/resourceMap';
@import './reservationCalendar/resourceReservationCalendar';
