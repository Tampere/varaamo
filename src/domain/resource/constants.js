export const resourceProductTypes = Object.freeze({
  RENT: 'rent',
  EXTRA: 'extra',
});

export const resourcePriceTypes = Object.freeze({
  HOURLY: 'hourly',
  DAILY: 'daily',
  WEEKLY: 'weekly',
  FIXED: 'fixed',
});
