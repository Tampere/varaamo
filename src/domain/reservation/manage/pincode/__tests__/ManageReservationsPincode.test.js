import React from 'react';
import toJSON from 'enzyme-to-json';
import { shallow } from 'enzyme';
import snakeCaseKeys from 'snakecase-keys';

import ManageReservationsPincode from '../ManageReservationsPincode';
import reservation from '../../../../../common/data/fixtures/reservation';
import resource from '../../../../../common/data/fixtures/resource';

// This project handles API responses differently based on the method
// that is used for fetching. Data in the redux store is in camelCase,
// but data fetched through the apiClient is in snake_case. This
// component was previously used in a context where API data
// originated from the redux store, but now lives in a context where
// this data comes directly from the apiClient.

// To be able to use the same test tooling, we are transforming the
// camelCase mock objects into snake_case mock objects.
const makeSnakeCase = obj => snakeCaseKeys(obj);

describe('ManageReservationsPincode', () => {
  function getWrapper(props) {
    return shallow(
      <ManageReservationsPincode {...props} />,
    );
  }

  const pendingCodeResource = resource.attr('generateAccessCodes', false);

  test('renders correctly', () => {
    const wrapper = getWrapper({ reservation: makeSnakeCase(reservation.build()) });
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  test('show Icon when accessCode pending', () => {
    const wrapper = getWrapper({
      reservation: reservation.build({ resource: makeSnakeCase(pendingCodeResource.build()) }),
    });
    expect(wrapper.debug()).toContain('img');
  });

  test('show **** when accessCode exist', () => {
    reservation.attr('accessCode', 1234);
    const wrapper = getWrapper({ reservation: makeSnakeCase(reservation.build()) });
    expect(wrapper.html()).toContain('****');
  });
});
