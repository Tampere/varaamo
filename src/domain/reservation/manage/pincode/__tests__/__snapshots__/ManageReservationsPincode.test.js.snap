// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageReservationsPincode renders correctly 1`] = `
<div
  className="app-ManageReservationPincode"
/>
`;
